File: src/graph.ts
import { allTags, type Note } from "./types.js";
import { scanVault, readNote } from "./vault.js";
import { stripFrontmatterLines } from "./markdown.js";
import {
  type FilterOptions,
  type PaginationOpts,
  type PaginatedResult,
  type CompiledFilter,
  paginate,
  ensureTrailingSlash,
  compileFilter,
  passesCompiledFilter,
  passesExcludePattern,
  filterNotes,
} from "./filter.js";

export interface GraphState {
  vaultPath: string;
  notes: Map<string, Note>;
  forward: Map<string, Set<string>>;
  backward: Map<string, Set<string>>;
  missing: Map<string, Set<string>>;
  fuzzyIndex: Map<string, string>;
}

export function buildGraph(vaultPath: string): GraphState {
  const paths = scanVault(vaultPath);
  const notes = new Map<string, Note>();

  for (const path of paths) {
    try {
      const note = readNote(path);
      notes.set(normalize(note.name), note);
    } catch {
      // skip unreadable files
    }
  }

  const { forward, backward, missing } = buildAdjacency(notes);

  const fuzzyIndex = new Map<string, string>();
  for (const [key, note] of notes) {
    const fk = normalizeFuzzy(note.name);
    if (!fuzzyIndex.has(fk)) fuzzyIndex.set(fk, key);
  }

  return { vaultPath, notes, forward, backward, missing, fuzzyIndex };
}

export function rebuildGraph(state: GraphState): void {
  const fresh = buildGraph(state.vaultPath);
  state.notes = fresh.notes;
  state.forward = fresh.forward;
  state.backward = fresh.backward;
  state.missing = fresh.missing;
  state.fuzzyIndex = fresh.fuzzyIndex;
}

function buildAdjacency(notes: Map<string, Note>) {
  const forward = new Map<string, Set<string>>();
  const backward = new Map<string, Set<string>>();
  const missing = new Map<string, Set<string>>();

  for (const [sourceKey, note] of notes) {
    const targets = new Set(note.wikilinks.map((wl) => normalize(wl.name)));
    forward.set(sourceKey, targets);

    for (const targetKey of targets) {
      if (!backward.has(targetKey)) backward.set(targetKey, new Set());
      backward.get(targetKey)!.add(sourceKey);

      if (!notes.has(targetKey)) {
        if (!missing.has(targetKey)) missing.set(targetKey, new Set());
        missing.get(targetKey)!.add(sourceKey);
      }
    }
  }

  return { forward, backward, missing };
}

export function resolve(name: string, state: GraphState): Note | null {
  const key = normalize(name);
  const note = state.notes.get(key);
  if (note) return note;
  return fuzzyResolve(name, state);
}

function fuzzyResolve(name: string, state: GraphState): Note | null {
  const fk = normalizeFuzzy(name);
  const key = state.fuzzyIndex.get(fk);
  if (key) return state.notes.get(key) ?? null;
  return null;
}

// --- Tools ---

export function batchResolve(names: string[], state: GraphState) {
  const resolved: {
    name: string;
    path: string;
    content: string;
    frontmatter: Record<string, unknown>;
    tags: string[];
    frontmatterTags: string[];
    inlineTags: string[];
    headings: Note["headings"];
    checkboxes: Note["checkboxes"];
  }[] = [];
  const errors: { name: string; error: string }[] = [];

  for (const name of names) {
    const note = resolve(name, state);
    if (note) {
      resolved.push({
        name: note.name,
        path: note.path,
        content: note.content,
        frontmatter: note.frontmatter ?? {},
        tags: allTags(note),
        frontmatterTags: note.frontmatterTags,
        inlineTags: note.inlineTags,
        headings: note.headings,
        checkboxes: note.checkboxes,
      });
    } else {
      errors.push({ name, error: `Note '${name}' not found` });
    }
  }

  return { resolved, errors };
}

function filterBacklinkRef(
  refKey: string,
  state: GraphState,
  folder: string | undefined,
  cf: CompiledFilter,
): { name: string; path: string | null } | null {
  const n = state.notes.get(refKey);
  const name = n ? n.name : refKey;
  const path = n ? n.path : null;

  if (folder) {
    if (!path) return { name, path };
    const relPath = path.slice(state.vaultPath.length + 1);
    if (!relPath.startsWith(ensureTrailingSlash(folder))) return null;
  }

  if (!n) {
    if (!passesExcludePattern(name, cf.excludeRe)) return null;
  } else {
    if (!passesCompiledFilter(n, state.vaultPath, cf)) return null;
  }
  return { name, path };
}

export function backlinks(
  noteName: string,
  state: GraphState,
  opts: PaginationOpts & FilterOptions = {},
): PaginatedResult<{ name: string; path: string | null }> {
  const note = resolve(noteName, state);
  if (!note) return paginate([], opts);
  const key = normalize(note.name);
  const refs = state.backward.get(key);
  if (!refs) return paginate([], opts);

  const cf = compileFilter({
    excludeFolders: opts.excludeFolders,
    excludePattern: opts.excludePattern,
    modifiedAfter: opts.modifiedAfter,
    modifiedBefore: opts.modifiedBefore,
    tags: opts.tags,
    excludeTags: opts.excludeTags,
    tagsMode: opts.tagsMode,
  });

  const all: { name: string; path: string | null }[] = [];
  for (const refKey of refs) {
    const result = filterBacklinkRef(refKey, state, opts.folder, cf);
    if (result) all.push(result);
  }
  all.sort((a, b) => a.name.localeCompare(b.name));

  return paginate(all, opts);
}

export function batchFindBacklinks(
  names: string[],
  state: GraphState,
  opts: PaginationOpts & FilterOptions = {},
) {
  const results: {
    name: string;
    backlinks: PaginatedResult<{ name: string; path: string | null }>;
  }[] = [];
  const errors: { name: string; error: string }[] = [];

  for (const name of names) {
    const note = resolve(name, state);
    if (note) {
      results.push({
        name: note.name,
        backlinks: backlinks(note.name, state, opts),
      });
    } else {
      errors.push({ name, error: `Note '${name}' not found` });
    }
  }

  return { results, errors };
}

export function orphans(
  state: GraphState,
  opts: PaginationOpts & FilterOptions = {},
): PaginatedResult<{
  name: string;
  path: string;
  tags: string[];
  frontmatterTags: string[];
  inlineTags: string[];
  content_length: number;
  empty: boolean;
}> {
  const cf = compileFilter(opts);

  // If excludePattern was provided but didn't compile, return empty (matching existing convention)
  if (opts.excludePattern && !cf.excludeRe) {
    return {
      total: 0,
      offset: opts.offset ?? 0,
      limit: opts.limit ?? 50,
      results: [],
    };
  }

  const orphanList: {
    name: string;
    path: string;
    tags: string[];
    frontmatterTags: string[];
    inlineTags: string[];
    content_length: number;
    empty: boolean;
  }[] = [];
  for (const [key, note] of state.notes) {
    if (!passesCompiledFilter(note, state.vaultPath, cf)) continue;
    const refs = state.backward.get(key);
    if (!refs || refs.size === 0) {
      const bodyLines = stripFrontmatterLines(note.content.split(/\r?\n/));
      const bodyText = bodyLines.join("\n").trim();
      orphanList.push({
        name: note.name,
        path: note.path,
        tags: allTags(note),
        frontmatterTags: note.frontmatterTags,
        inlineTags: note.inlineTags,
        content_length: note.content.length,
        empty: bodyText.length === 0,
      });
    }
  }
  orphanList.sort((a, b) => a.name.localeCompare(b.name));

  return paginate(orphanList, opts);
}

type BrokenLinkType = "note" | "embed" | "all";

const HAS_EXTENSION = /\.\w{1,5}$/;

function passesTypeFilter(name: string, type: BrokenLinkType): boolean {
  if (type === "all") return true;
  const isEmbed = HAS_EXTENSION.test(name);
  return type === "embed" ? isEmbed : !isEmbed;
}

function buildScopeKeys(
  noteNames: string[] | undefined,
  state: GraphState,
): Set<string> | null {
  if (!noteNames) return null;
  const keys = new Set<string>();
  for (const n of noteNames) {
    const resolved = resolve(n, state);
    if (resolved) keys.add(normalize(resolved.name));
  }
  return keys;
}

function filterMissingEntry(
  name: string,
  referrers: Set<string>,
  type: BrokenLinkType,
  scopeKeys: Set<string> | null,
  state: GraphState,
  cf: CompiledFilter,
): { name: string; referenced_by: string[]; count: number } | null {
  if (!passesTypeFilter(name, type)) return null;

  const filteredReferrers = [...referrers].filter((refKey) => {
    if (scopeKeys && !scopeKeys.has(refKey)) return false;
    const note = state.notes.get(refKey);
    if (!note) return false;
    return passesCompiledFilter(note, state.vaultPath, cf);
  });
  if (filteredReferrers.length === 0) return null;

  return {
    name,
    referenced_by: filteredReferrers,
    count: filteredReferrers.length,
  };
}

export function missingNotes(
  state: GraphState,
  opts: PaginationOpts &
    FilterOptions & {
      type?: BrokenLinkType;
      noteNames?: string[];
    } = {},
): PaginatedResult<{ name: string; referenced_by: string[]; count: number }> {
  const type = opts.type ?? "all";
  const cf = compileFilter(opts);
  const scopeKeys = buildScopeKeys(opts.noteNames, state);

  const missingList: {
    name: string;
    referenced_by: string[];
    count: number;
  }[] = [];
  for (const [name, referrers] of state.missing) {
    const entry = filterMissingEntry(
      name,
      referrers,
      type,
      scopeKeys,
      state,
      cf,
    );
    if (entry) missingList.push(entry);
  }
  missingList.sort((a, b) => b.count - a.count);

  return paginate(missingList, opts);
}

function isReferrerInScope(
  refKey: string,
  scopeKeys: Set<string> | null,
  state: GraphState,
  cf: CompiledFilter,
): boolean {
  if (scopeKeys && !scopeKeys.has(refKey)) return false;
  const note = state.notes.get(refKey);
  if (!note) return false;
  return passesCompiledFilter(note, state.vaultPath, cf);
}

function collectBySource(
  state: GraphState,
  type: BrokenLinkType,
  scopeKeys: Set<string> | null,
  cf: CompiledFilter,
): Map<string, { name: string; is_embed: boolean }[]> {
  const bySource = new Map<string, { name: string; is_embed: boolean }[]>();

  for (const [name, referrers] of state.missing) {
    if (!passesTypeFilter(name, type)) continue;
    const isEmbed = HAS_EXTENSION.test(name);

    for (const refKey of referrers) {
      if (!isReferrerInScope(refKey, scopeKeys, state, cf)) continue;
      if (!bySource.has(refKey)) bySource.set(refKey, []);
      bySource.get(refKey)!.push({ name, is_embed: isEmbed });
    }
  }

  return bySource;
}

export function missingNotesBySource(
  state: GraphState,
  opts: PaginationOpts &
    FilterOptions & {
      type?: BrokenLinkType;
      noteNames?: string[];
    } = {},
): PaginatedResult<{
  source: string;
  source_path: string;
  broken_links: { name: string; is_embed: boolean }[];
  count: number;
}> {
  const type = opts.type ?? "all";
  const cf = compileFilter(opts);
  const scopeKeys = buildScopeKeys(opts.noteNames, state);
  const bySource = collectBySource(state, type, scopeKeys, cf);

  const sourceList = [...bySource.entries()]
    .map(([key, broken_links]) => {
      const note = state.notes.get(key)!;
      return {
        source: note.name,
        source_path: note.path,
        broken_links,
        count: broken_links.length,
      };
    })
    .sort((a, b) => b.count - a.count || a.source.localeCompare(b.source));

  return paginate(sourceList, opts);
}

// eslint-disable-next-line sonarjs/cognitive-complexity -- BFS neighbor processing with exclude/link-type detection
function processNeighbors(
  sourceKey: string,
  depth: number,
  state: GraphState,
  excludePrefixes: string[],
  linkTypes: Map<string, "wikilink" | "embed">,
  missingFound: Set<string>,
  queue: [string, number][],
): void {
  const sourceNote = state.notes.get(sourceKey);
  if (!sourceNote) return;

  const targets = state.forward.get(sourceKey);
  if (!targets) return;

  for (const target of targets) {
    if (!state.notes.has(target)) {
      missingFound.add(target);
      continue;
    }

    if (excludePrefixes.length > 0) {
      const targetNote = state.notes.get(target)!;
      const relPath = targetNote.path.slice(state.vaultPath.length + 1);
      if (excludePrefixes.some((p) => relPath.startsWith(p))) continue;
    }

    if (!linkTypes.has(target)) {
      const wl = sourceNote.wikilinks.find((w) => normalize(w.name) === target);
      if (wl) {
        linkTypes.set(target, wl.embed ? "embed" : "wikilink");
      }
    }

    queue.push([target, depth + 1]);
  }
}

export function traverse(
  noteName: string | string[],
  maxDepth: number,
  state: GraphState,
  opts: { excludeFolders?: string[] } = {},
) {
  const names = Array.isArray(noteName) ? noteName : [noteName];
  const isMultiRoot = Array.isArray(noteName);

  const startNotes: Note[] = [];
  for (const n of names) {
    const note = resolve(n, state);
    if (!note) {
      return {
        error: `Note '${n}' not found`,
        notes: [],
        missing: [],
      };
    }
    startNotes.push(note);
  }

  const excludePrefixes = (opts.excludeFolders ?? []).map(ensureTrailingSlash);

  const visited = new Map<string, number>();
  const missingFound = new Set<string>();
  const linkTypes = new Map<string, "wikilink" | "embed">();

  // BFS
  const queue: [string, number][] = [];
  for (const note of startNotes) {
    const key = normalize(note.name);
    queue.push([key, 0]);
  }

  let head = 0;
  while (head < queue.length) {
    const [key, depth] = queue[head++];
    if (visited.has(key)) continue;
    visited.set(key, depth);

    if (depth < maxDepth) {
      processNeighbors(
        key,
        depth,
        state,
        excludePrefixes,
        linkTypes,
        missingFound,
        queue,
      );
    }
  }

  const notes = [...visited.entries()]
    .map(([key, depth]) => {
      const note = state.notes.get(key)!;
      return {
        name: note.name,
        path: note.path,
        depth,
        frontmatter: note.frontmatter,
        tags: allTags(note),
        frontmatterTags: note.frontmatterTags,
        inlineTags: note.inlineTags,
        link_type: (linkTypes.get(key) as "wikilink" | "embed" | null) ?? null,
      };
    })
    .sort((a, b) => a.depth - b.depth || a.name.localeCompare(b.name));

  const missingResult = [...missingFound].map((name) => {
    const referrers = state.missing.get(name);
    return {
      name,
      referenced_by: referrers ? [...referrers] : [],
    };
  });

  return {
    ...(isMultiRoot
      ? { roots: startNotes.map((n) => n.name) }
      : { root: startNotes[0].name }),
    depth: maxDepth,
    notes,
    missing: missingResult,
  };
}

export function stats(state: GraphState, opts: FilterOptions = {}) {
  const inScope = filterNotes(state.notes, state.vaultPath, opts);

  let tagged = 0;
  let orphanCount = 0;
  for (const [key, note] of inScope) {
    if (note.frontmatterTags.length > 0 || note.inlineTags.length > 0) tagged++;
    const refs = state.backward.get(key);
    if (!refs || refs.size === 0) orphanCount++;
  }

  // Count missing links referenced only by in-scope notes
  let missingCount = 0;
  for (const [, referrers] of state.missing) {
    if ([...referrers].some((refKey) => inScope.has(refKey))) missingCount++;
  }

  return {
    total_notes: inScope.size,
    tagged,
    untagged: inScope.size - tagged,
    orphans: orphanCount,
    missing_links: missingCount,
  };
}

function normalize(name: string): string {
  return name.trim().toLowerCase();
}

function normalizeFuzzy(name: string): string {
  return (
    name
      .trim()
      .toLowerCase()
      .replace(/[-_]/g, "")
      .normalize("NFD")
      // Keep only printable ASCII for fuzzy matching (strips diacritics, control chars, etc.)
      .replace(/[^\u0020-\u007E]/g, "")
  );
}

File: lib/vault_kit/graph.ex
defmodule VaultKit.Graph do
  @moduledoc """
  GenServer holding an in-memory link graph of the vault.

  Built on startup by scanning all notes and constructing forward/backward
  adjacency maps for instant link resolution.
  """

  use GenServer

  alias VaultKit.{Note, Vault}

  @type state :: %{
          vault_path: String.t(),
          notes: %{String.t() => Note.t()},
          forward: %{String.t() => MapSet.t()},
          backward: %{String.t() => MapSet.t()},
          missing: %{String.t() => MapSet.t()}
        }

  # Client API

  def start_link(opts) do
    vault_path = Keyword.fetch!(opts, :vault_path)
    GenServer.start_link(__MODULE__, vault_path, name: __MODULE__)
  end

  @doc "BFS traversal from a note to depth N. Returns notes + content + missing links."
  @spec traverse(String.t(), non_neg_integer(), keyword()) :: map()
  def traverse(note_name, depth \\ 2, opts \\ []) do
    GenServer.call(__MODULE__, {:traverse, note_name, depth, opts}, 30_000)
  end

  @doc "Returns notes that link to the given note."
  @spec backlinks(String.t()) :: [map()]
  def backlinks(note_name) do
    GenServer.call(__MODULE__, {:backlinks, note_name})
  end

  @doc "Returns notes with no incoming links."
  @spec orphans(keyword()) :: map()
  def orphans(opts \\ []) do
    GenServer.call(__MODULE__, {:orphans, opts}, 30_000)
  end

  @doc "Returns all broken links with reference counts."
  @spec missing_notes(keyword()) :: map()
  def missing_notes(opts \\ []) do
    GenServer.call(__MODULE__, {:missing_notes, opts}, 30_000)
  end

  @doc "Case-insensitive note lookup with fuzzy fallback."
  @spec resolve(String.t()) :: {:ok, Note.t()} | :not_found
  def resolve(name) do
    GenServer.call(__MODULE__, {:resolve, name})
  end

  @doc "Rebuild the graph from disk."
  @spec rebuild() :: :ok
  def rebuild do
    GenServer.call(__MODULE__, :rebuild, 60_000)
  end

  @doc "Return vault statistics."
  @spec stats() :: map()
  def stats do
    GenServer.call(__MODULE__, :stats, 30_000)
  end

  @doc "Search vault content for a query string."
  @spec search(String.t(), keyword()) :: map()
  def search(query, opts \\ []) do
    GenServer.call(__MODULE__, {:search, query, opts}, 30_000)
  end

  @doc "Return the vault path."
  @spec vault_path() :: String.t()
  def vault_path do
    GenServer.call(__MODULE__, :get_vault_path)
  end

  # Server callbacks

  @impl true
  def init(vault_path) do
    state = build_graph(vault_path)
    {:ok, state}
  end

  @impl true
  def handle_call({:traverse, note_name, depth, _opts}, _from, state) do
    result = do_traverse(note_name, depth, state)
    {:reply, result, state}
  end

  def handle_call({:backlinks, note_name}, _from, state) do
    key = normalize(note_name)

    refs =
      state.backward
      |> Map.get(key, MapSet.new())
      |> MapSet.to_list()
      |> Enum.map(fn ref_key ->
        case Map.get(state.notes, ref_key) do
          nil -> %{name: ref_key, path: nil}
          note -> %{name: note.name, path: note.path}
        end
      end)

    {:reply, refs, state}
  end

  def handle_call({:orphans, opts}, _from, state) do
    limit = Keyword.get(opts, :limit, 50)
    offset = Keyword.get(opts, :offset, 0)

    orphan_list =
      state.notes
      |> Enum.filter(fn {key, _note} ->
        refs = Map.get(state.backward, key, MapSet.new())
        MapSet.size(refs) == 0
      end)
      |> Enum.map(fn {_key, note} ->
        %{name: note.name, path: note.path, tags: note.tags}
      end)
      |> Enum.sort_by(& &1.name)

    total = length(orphan_list)
    page = orphan_list |> Enum.drop(offset) |> Enum.take(limit)

    {:reply, %{total: total, offset: offset, limit: limit, results: page}, state}
  end

  def handle_call({:missing_notes, opts}, _from, state) do
    limit = Keyword.get(opts, :limit, 50)
    offset = Keyword.get(opts, :offset, 0)

    missing_list =
      state.missing
      |> Enum.map(fn {name, referrers} ->
        %{name: name, referenced_by: MapSet.to_list(referrers), count: MapSet.size(referrers)}
      end)
      |> Enum.sort_by(& &1.count, :desc)

    total = length(missing_list)
    page = missing_list |> Enum.drop(offset) |> Enum.take(limit)

    {:reply, %{total: total, offset: offset, limit: limit, results: page}, state}
  end

  def handle_call({:resolve, name}, _from, state) do
    result = do_resolve(name, state)
    {:reply, result, state}
  end

  def handle_call(:rebuild, _from, state) do
    new_state = build_graph(state.vault_path)
    {:reply, :ok, new_state}
  end

  def handle_call(:get_vault_path, _from, state) do
    {:reply, state.vault_path, state}
  end

  def handle_call({:search, query, opts}, _from, state) do
    limit = Keyword.get(opts, :limit, 50)
    offset = Keyword.get(opts, :offset, 0)
    q = String.downcase(query)
    needed = offset + limit

    {total, collected} =
      Enum.reduce(state.notes, {0, []}, fn {_key, note}, {count, acc} ->
        matches =
          note.content
          |> String.split(~r/\r?\n/)
          |> Enum.with_index(1)
          |> Enum.filter(fn {line, _} -> String.contains?(String.downcase(line), q) end)

        new_count = count + length(matches)

        new_acc =
          if length(acc) < needed do
            remaining = needed - length(acc)

            extra =
              matches
              |> Enum.take(remaining)
              |> Enum.map(fn {line, line_num} ->
                %{file: note.name, path: note.path, line: line_num, text: String.trim(line)}
              end)

            acc ++ extra
          else
            acc
          end

        {new_count, new_acc}
      end)

    page = collected |> Enum.drop(offset) |> Enum.take(limit)

    {:reply, %{total: total, offset: offset, limit: limit, results: page}, state}
  end

  def handle_call(:stats, _from, state) do
    total = map_size(state.notes)

    tagged =
      state.notes
      |> Enum.count(fn {_, note} -> note.tags != [] end)

    orphan_count =
      state.notes
      |> Enum.count(fn {key, _} ->
        refs = Map.get(state.backward, key, MapSet.new())
        MapSet.size(refs) == 0
      end)

    result = %{
      total_notes: total,
      tagged: tagged,
      untagged: total - tagged,
      orphans: orphan_count,
      missing_links: map_size(state.missing)
    }

    {:reply, result, state}
  end

  # Private

  defp build_graph(vault_path) do
    paths = Vault.scan(vault_path)

    notes =
      paths
      |> Enum.reduce(%{}, fn path, acc ->
        case Vault.read_note(path) do
          {:ok, note} -> Map.put(acc, normalize(note.name), note)
          {:error, _} -> acc
        end
      end)

    {forward, backward, missing} = build_adjacency(notes)

    %{
      vault_path: vault_path,
      notes: notes,
      forward: forward,
      backward: backward,
      missing: missing
    }
  end

  defp build_adjacency(notes) do
    Enum.reduce(notes, {%{}, %{}, %{}}, fn {source_key, note}, {fwd, bwd, miss} ->
      targets =
        note.wikilinks
        |> Enum.map(fn wl -> normalize(wl.name) end)
        |> Enum.uniq()

      new_fwd = Map.put(fwd, source_key, MapSet.new(targets))

      {new_bwd, new_miss} =
        Enum.reduce(targets, {bwd, miss}, fn target_key, {b, m} ->
          new_b = Map.update(b, target_key, MapSet.new([source_key]), &MapSet.put(&1, source_key))

          new_m =
            if Map.has_key?(notes, target_key) do
              m
            else
              Map.update(m, target_key, MapSet.new([source_key]), &MapSet.put(&1, source_key))
            end

          {new_b, new_m}
        end)

      {new_fwd, new_bwd, new_miss}
    end)
  end

  defp do_traverse(note_name, max_depth, state) do
    start_key = normalize(note_name)

    case do_resolve(note_name, state) do
      {:ok, start_note} ->
        queue = :queue.from_list([{start_key, 0}])
        {visited, missing_found} = bfs(queue, max_depth, state, %{}, MapSet.new())

        notes =
          visited
          |> Enum.map(fn {key, depth} ->
            note = Map.get(state.notes, key)

            %{
              name: note.name,
              path: note.path,
              depth: depth,
              frontmatter: note.frontmatter,
              tags: note.tags
            }
          end)
          |> Enum.sort_by(&{&1.depth, &1.name})

        missing =
          missing_found
          |> MapSet.to_list()
          |> Enum.map(fn name ->
            referrers = Map.get(state.missing, name, MapSet.new())
            %{name: name, referenced_by: MapSet.to_list(referrers)}
          end)

        %{root: start_note.name, depth: max_depth, notes: notes, missing: missing}

      :not_found ->
        %{error: "Note '#{note_name}' not found", notes: [], missing: []}
    end
  end

  defp bfs(queue, max_depth, state, visited, missing) do
    case :queue.out(queue) do
      {:empty, _} ->
        {visited, missing}

      {{:value, {key, depth}}, rest} ->
        if Map.has_key?(visited, key) do
          bfs(rest, max_depth, state, visited, missing)
        else
          new_visited = Map.put(visited, key, depth)

          if depth < max_depth do
            targets = Map.get(state.forward, key, MapSet.new()) |> MapSet.to_list()
            {new_queue, new_missing} = enqueue_targets(targets, depth, state, rest, missing)
            bfs(new_queue, max_depth, state, new_visited, new_missing)
          else
            bfs(rest, max_depth, state, new_visited, missing)
          end
        end
    end
  end

  defp enqueue_targets(targets, depth, state, queue, missing) do
    Enum.reduce(targets, {queue, missing}, fn target, {q, m} ->
      if Map.has_key?(state.notes, target) do
        {:queue.in({target, depth + 1}, q), m}
      else
        {q, MapSet.put(m, target)}
      end
    end)
  end

  defp do_resolve(name, state) do
    key = normalize(name)

    case Map.get(state.notes, key) do
      nil -> fuzzy_resolve(name, state)
      note -> {:ok, note}
    end
  end

  defp fuzzy_resolve(name, state) do
    target = normalize_fuzzy(name)

    result =
      state.notes
      |> Enum.find(fn {_key, note} ->
        normalize_fuzzy(note.name) == target
      end)

    case result do
      {_key, note} -> {:ok, note}
      nil -> :not_found
    end
  end

  defp normalize(name) do
    name |> String.trim() |> String.downcase()
  end

  defp normalize_fuzzy(name) do
    name
    |> String.trim()
    |> String.downcase()
    |> String.replace(~r/[-_]/, "")
    |> String.normalize(:nfd)
    |> String.replace(~r/[^\x00-\x7F]/, "")
  end
end
